core/src/main/kotlin/com/github/quillraven/quillysadventure/Main.kt

```diff
@@ -127,50 +130,57 @@ class Main(
         // init Box2D - the next call avoids some issues with older devices where the box2d libraries were not loaded correctly
         Box2D.init()
 
-        // setup context and register stuff that should also be disposed at the end of the game lifecycle
-        ctx.register {
-            bindSingleton(ShaderPrograms())
-            bindSingleton(SpriteBatch(2048))
-            bindSingleton(AssetManager().apply {
-                // we use tmx tiled maps created via the Tiled tool and therefore
-                // we use the TmxMapLoader for our assetmanager to be able to
-                // load/unload .tmx files
-                setLoader(TiledMap::class.java, TmxMapLoader(fileHandleResolver))
-            })
-            bindSingleton(Stage(FitViewport(VIRTUAL_W.toFloat(), VIRTUAL_H.toFloat()), ctx.inject<SpriteBatch>()))
-            bindSingleton(createSkin(ctx.inject()))
-            bindSingleton(RayHandler(world))
-            bindSingleton(Box2DDebugRenderer())
-            bindSingleton(OrthogonalTiledMapRenderer(null, UNIT_SCALE, ctx.inject<SpriteBatch>()))
+        KtxAsync.initiate()
+        val assetStorage = AssetStorage().apply {
+            // we use tmx tiled maps created via the Tiled tool and therefore
+            // we use the TmxMapLoader to be able to load/unload .tmx files
+            setLoader<TiledMap> { TmxMapLoader(this.fileResolver) }
         }
-
-        // we need a multiplexer to react on the following input events
-        // UI widget --> Stage
-        // keyboard --> InputProcessor (GameEventManager)
-        Gdx.input.inputProcessor = InputMultiplexer(gameEventManager, ctx.inject<Stage>())
-
-        // box2d light should not create shadows for dynamic game objects
-        Light.setGlobalContactFilter(FILTER_CATEGORY_LIGHT, 0, FILTER_MASK_LIGHTS)
-
-        // initial screen is the loading screen which is loading all assets for the game
-        addScreen(
-            LoadingScreen(
-                this, // game instance to switch screens
-                ctx.inject<AssetManager>()[I18nAssets.DEFAULT],
-                ctx.inject(), // stage
-                ctx.inject(), // assets
-                gameEventManager, // game event manager
-                audioService,
-                world, // physic world
-                ecsEngine, // entity component engine
-                ctx.inject(), // ray handler
-                ctx.inject(), // shaders
-                ctx.inject(), // sprite batch
-                ctx.inject(), // tiled map renderer
-                ctx.inject() // box2d debug renderer
+        KtxAsync.launch {
+            // load textures for skin
+            val uiAtlas = assetStorage.loadAsync(TextureAtlasAssets.UI)
+            val bundle = assetStorage.loadAsync(I18nAssets.DEFAULT)
```

Comment:
This isn't a bad solution and you get the benefit of running the initialization code in parallel - asynchronous coroutines are loading your assets while you create the context, setup input processor etc.

However, this made me realize that something like `loadSync` that would block the thread until an asset is loaded might be useful for initialization of loading screen assets. This seems to be a common use case and is easy to achieve with `AssetManager` through `finishLoading`.

You can leave it as is (and I encourage you to, this isn't a bad way to init the app), but just so you know - I plan on adding synchronous loading to `AssetStorage` as well.